### PR TITLE
SQL: Implement `::` cast operator

### LINKED ISSUE #38717: SQL: Implement postgres compatible `::` cast operator
PostgreSQL has an alternative way to cast fields/literals/functions to another data type apart from `CAST <field/literal/function_call> AS <data_type>`, which is the `::` operator: `<field/literal/function_call>::<data_type>. It would be nice to implement this as it enhances user experience and can result in more compact queries.

### PR DESCRIPTION
**Pull request by matriv**
`<expression>::<dataType>` is a simplified altenative syntax to
`CAST(<expression> AS <dataType> which exists in PostgreSQL and
provides an improved user experience and possibly more compact
SQL queries.

Fixes: #38717

### CODE REVIEW

File: x-pack/plugin/sql/src/main/antlr/SqlBase.g4 [hunk 1]
```diff
@@ -224,22 +225,22 @@ primaryExpression
     | '(' expression ')'                                                             #parenthesizedExpression
     ;
 
+builtinDateTimeFunction
```
Comment (costin) [inline x-pack/plugin/sql/src/main/antlr/SqlBase.g4]: Why was the order changed?

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/parser/ExpressionTests.java [hunk 2]
```diff
@@ -282,6 +282,37 @@ public void testCastWithInvalidDataType() {
...
+        assertEquals(DataType.LONG, cast.dataType());
+        assertEquals(Mul.class, cast.field().getClass());
+        Mul mul = (Mul) cast.field();
+        assertEquals("10*2", mul.sourceText());
+        assertEquals(DataType.INTEGER, mul.dataType());
+    }
+
+    public void testCastOperatorWithInvalidDataType() {
+        ParsingException ex = expectThrows(ParsingException.class, () -> parser.createExpression("-1::INVALID"));
+        assertEquals("line 1:6: Does not recognize type invalid", ex.getMessage());
```
Comment (costin) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/parser/ExpressionTests.java]: The message needs improving - use "Does not recognize type [{}]", type.sourceText() - to preserve the original declaration inside visitPrimitiveDataType

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/src/main/antlr/SqlBase.g4 (hunk 1 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/antlr/SqlBase.g4]: just to have the cast/convert declarations next to each other.

**Review by astefan — APPROVED** [review]
LGTM

**Comment by matriv** [comment]
Backported to `7.x` with 032bcf99d62dac0d219f43cd7350cde314ef51fd

**Pull request merged**